File: app/src/test/java/dev/bee/kanjianki/SettingsLearningStepsPanelModelsTest.java
package dev.bee.kanjianki;

import dev.bee.kanjianki.core.RecordsSchedulerModels;

import org.junit.Test;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;

public final class SettingsLearningStepsPanelModelsTest {
    @Test
    public void createBuildsLearningStepsPanelModelFromCurrentSettings() {
        RecordsSchedulerModels.LearningStepSettings current = new RecordsSchedulerModels.LearningStepSettings(
                Arrays.asList(2, 15),
                Arrays.asList(7)
        );
        RecordingSaveAction action = new RecordingSaveAction();

        SettingsLearningStepsPanelModel model = SettingsLearningStepsPanelModels.create(current, action);

        assertEquals("Learning steps", model.getTitle());
        assertEquals("New cards", model.getNewCardsLabel());
        assertEquals("2m, 15m", model.getInitialNewStepsText());
        assertEquals("7m", model.getInitialReviewStepsText());
        assertEquals("1m, 10m", model.getDefaultNewStepsText());
        assertEquals("10m", model.getDefaultReviewStepsText());

        model.getOnSave().save("3m", "8m");
        assertEquals("3m", action.newStepsText);
        assertEquals("8m", action.reviewStepsText);
    }

    private static final class RecordingSaveAction implements SettingsLearningStepsSaveAction {
        String newStepsText;
        String reviewStepsText;

        @Override
        public void save(String newStepsText, String reviewStepsText) {
            this.newStepsText = newStepsText;
            this.reviewStepsText = reviewStepsText;
        }
    }
}

File: app/src/test/kotlin/dev/bee/kanjianki/SettingsLearningStepsPanelModelsTest.kt
package dev.bee.kanjianki

import dev.bee.kanjianki.core.RecordsSchedulerModels
import org.junit.Assert.assertEquals
import org.junit.Test

class SettingsLearningStepsPanelModelsTest {
    @Test
    fun createBuildsLearningStepsPanelModelFromCurrentSettings() {
        val current = RecordsSchedulerModels.LearningStepSettings(
            listOf(2, 15),
            listOf(7),
        )
        var newStepsText: String? = null
        var reviewStepsText: String? = null

        val model = SettingsLearningStepsPanelModels.create(current) { newStepsTextArg, reviewStepsTextArg ->
            newStepsText = newStepsTextArg
            reviewStepsText = reviewStepsTextArg
        }

        assertEquals("Learning steps", model.title)
        assertEquals("New cards", model.newCardsLabel)
        assertEquals("2m, 15m", model.initialNewStepsText)
        assertEquals("7m", model.initialReviewStepsText)
        assertEquals("1m, 10m", model.defaultNewStepsText)
        assertEquals("10m", model.defaultReviewStepsText)

        model.onSave.save("3m", "8m")
        assertEquals("3m", newStepsText)
        assertEquals("8m", reviewStepsText)
    }

    @Test
    fun createKeepsBlankReviewStepsVisibleAndSavable() {
        val current = RecordsSchedulerModels.LearningStepSettings(
            listOf(1, 10),
            emptyList(),
        )
        var newStepsText: String? = null
        var reviewStepsText: String? = null

        val model = SettingsLearningStepsPanelModels.create(current) { newStepsTextArg, reviewStepsTextArg ->
            newStepsText = newStepsTextArg
            reviewStepsText = reviewStepsTextArg
        }

        assertEquals("1m, 10m", model.initialNewStepsText)
        assertEquals("", model.initialReviewStepsText)

        model.onSave.save("1m, 10m", "")
        assertEquals("1m, 10m", newStepsText)
        assertEquals("", reviewStepsText)
    }

    @Test
    fun useNewCardStepsUsesCurrentNewCardTextInsteadOfResettingToDefaults() {
        val textState = SettingsLearningStepsPanelModels.useNewCardStepsTextState("4m, 30m")

        assertEquals("4m, 30m", textState.newStepsText)
        assertEquals("4m, 30m", textState.reviewStepsText)
    }
}
